File: app/argv.ts
import { CommandError } from './napi'

export type LaunchAction =
  | { kind: 'open' }
  | { kind: 'extract-here' }
  | { kind: 'extract-to'; destDir: string | null }
  | { kind: 'index-only' }

export type LaunchIntent = {
  action: LaunchAction
  archives: string[]
  silent: boolean
}

type ActionKind = 'open' | 'extract-here' | 'extract-to' | 'index-only'

export function launchArgsFromProcess(argv: string[]): string[] {
  const rest = argv.slice(1)
  const first = rest[0]
  if (first == null) {
    return []
  }
  const base = first.replaceAll('\\', '/').split('/').pop() ?? ''
  if (base.endsWith('.tsx') || base.endsWith('.ts') || base.endsWith('.js') || base.endsWith('.mjs')) {
    return rest.slice(1)
  }
  return rest
}

export function parseLaunchArgv(args: string[]): LaunchIntent {
  let silent = false
  let action: ActionKind | null = null
  let extractToOmitted = false
  const positionals: string[] = []
  let i = 0
  while (i < args.length) {
    const a = args[i] ?? ''
    if (a === '--silent') {
      silent = true
    } else if (a === '--extract-here') {
      action = setAction(action, 'extract-here')
    } else if (a === '--index-only') {
      action = setAction(action, 'index-only')
    } else if (a === '--extract-to') {
      action = setAction(action, 'extract-to')
      const next = args[i + 1]
      if (next == null || next === '--') {
        extractToOmitted = true
        if (next === '--') {
          i += 1
        }
      } else if (next.startsWith('--')) {
        extractToOmitted = true
      }
    } else if (a === '--') {
      positionals.push(...args.slice(i + 1))
      break
    } else if (a.startsWith('-') && a !== '-') {
      throw new CommandError('Internal', `unknown option '${a}'`, false)
    } else {
      positionals.push(a)
    }
    i += 1
  }

  let resolved: LaunchAction
  if (action === 'extract-here') {
    resolved = { kind: 'extract-here' }
  } else if (action === 'index-only') {
    resolved = { kind: 'index-only' }
  } else if (action === 'extract-to') {
    // One remaining path is the archive, never destDir (`--extract-to -- "%1"`).
    if (extractToOmitted || positionals.length <= 1) {
      resolved = { kind: 'extract-to', destDir: null }
    } else {
      const destDir = positionals.shift() ?? null
      resolved = { kind: 'extract-to', destDir }
    }
  } else {
    resolved = { kind: 'open' }
  }
  return { action: resolved, archives: positionals, silent }
}

function setAction(current: ActionKind | null, next: ActionKind): ActionKind {
  if (current == null || current === 'open' || current === next) {
    return next
  }
  throw new CommandError('Internal', `conflicting actions ${current} and ${next}`, false)
}

export function extractHereDest(archive: string): string {
  const norm = archive.replaceAll('\\', '/')
  const idx = norm.lastIndexOf('/')
  if (idx <= 0) {
    return '.'
  }
  return archive.slice(0, idx)
}

export function isHeadlessLaunch(intent: LaunchIntent): boolean {
  return isHeadlessAction(intent.action.kind, intent.silent)
}

export function isHeadlessAction(action: string, silent: boolean): boolean {
  if (action === 'index-only') {
    return true
  }
  if (!silent) {
    return false
  }
  return action === 'extract-here' || action === 'extract-to'
}

export type ArgvVector = {
  args: string[]
  action: 'open' | 'extract-here' | 'extract-to' | 'index-only'
  destDir: string | null
  archives: string[]
  silent: boolean
}

export function parseArgvVectors(text: string): ArgvVector[] {
  const out: ArgvVector[] = []
  for (const raw of text.split('\n')) {
    const line = raw.trim()
    if (line === '' || line.startsWith('#')) {
      continue
    }
    const cols = raw.split('\t')
    if (cols.length < 5) {
      throw new CommandError('Internal', `invalid argv vector: ${line}`, false)
    }
    const dest = cols[2] ?? ''
    const actionRaw = cols[1] ?? ''
    const action: ArgvVector['action'] =
      actionRaw === 'extract-here' || actionRaw === 'extract-to' || actionRaw === 'index-only'
        ? actionRaw
        : 'open'
    out.push({
      args: (cols[0] ?? '').split('|').filter(Boolean),
      action,
      destDir: dest === '' ? null : dest,
      archives: (cols[3] ?? '').split('|').filter(Boolean),
      silent: cols[4]?.trim() === '1',
    })
  }
  return out
}

File: native/src/argv.rs
use std::path::Path;

use crate::error::{ApiError, Result};
use crate::types::{ConfigOverwrite, Overwrite};

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum LaunchAction {
    Open,
    ExtractHere,
    /// `dest_dir` is `None` when omitted (`--extract-to -- <archive>` or a single
    /// remaining path). The archive is never treated as the destination.
    ExtractTo {
        dest_dir: Option<String>,
    },
    IndexOnly,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct LaunchIntent {
    pub action: LaunchAction,
    pub archives: Vec<String>,
    pub silent: bool,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum ActionKind {
    Open,
    ExtractHere,
    ExtractTo,
    IndexOnly,
}

pub fn parse_argv<I, S>(args: I) -> Result<LaunchIntent>
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    let args: Vec<String> = args.into_iter().map(|s| s.as_ref().to_string()).collect();
    let mut silent = false;
    let mut action: Option<ActionKind> = None;
    let mut extract_to_omitted = false;
    let mut positionals = Vec::new();
    let mut i = 0;
    while i < args.len() {
        let a = args[i].as_str();
        match a {
            "--silent" => silent = true,
            "--extract-here" => set_action(&mut action, ActionKind::ExtractHere)?,
            "--index-only" => set_action(&mut action, ActionKind::IndexOnly)?,
            "--extract-to" => {
                set_action(&mut action, ActionKind::ExtractTo)?;
                match args.get(i + 1).map(String::as_str) {
                    None | Some("--") => {
                        extract_to_omitted = true;
                        if args.get(i + 1).map(String::as_str) == Some("--") {
                            i += 1;
                        }
                    }
                    Some(next) if next.starts_with("--") => {
                        extract_to_omitted = true;
                    }
                    Some(_) => {}
                }
            }
            "--" => {
                positionals.extend(args[i + 1..].iter().cloned());
                break;
            }
            s if s.starts_with('-') && s != "-" => {
                return Err(ApiError::internal(format!("unknown option '{s}'")));
            }
            s => positionals.push(s.to_string()),
        }
        i += 1;
    }

    let action = match action.unwrap_or(ActionKind::Open) {
        ActionKind::Open => LaunchAction::Open,
        ActionKind::ExtractHere => LaunchAction::ExtractHere,
        ActionKind::IndexOnly => LaunchAction::IndexOnly,
        ActionKind::ExtractTo => {
            // One remaining path is the archive. Treating it as destDir is the
            // Windows ExtractTo (`--extract-to -- "%1"`) bug class.
            if extract_to_omitted || positionals.len() <= 1 {
                LaunchAction::ExtractTo { dest_dir: None }
            } else {
                let dest_dir = positionals.remove(0);
                LaunchAction::ExtractTo {
                    dest_dir: Some(dest_dir),
                }
            }
        }
    };
    Ok(LaunchIntent {
        action,
        archives: positionals,
        silent,
    })
}

fn set_action(current: &mut Option<ActionKind>, next: ActionKind) -> Result<()> {
    match *current {
        None | Some(ActionKind::Open) => {
            *current = Some(next);
            Ok(())
        }
        Some(prev) if prev == next => Ok(()),
        Some(prev) => Err(ApiError::internal(format!(
            "conflicting actions {} and {}",
            action_kind_name(prev),
            action_kind_name(next)
        ))),
    }
}

fn action_kind_name(kind: ActionKind) -> &'static str {
    match kind {
        ActionKind::Open => "open",
        ActionKind::ExtractHere => "extract-here",
        ActionKind::ExtractTo => "extract-to",
        ActionKind::IndexOnly => "index-only",
    }
}

/// `--silent` always maps to skip (never replace, never a hidden dialog).
/// Config `ask` also maps to skip on the headless path so native extract is never `'ask'`.
pub fn native_overwrite_for_launch(silent: bool, configured: ConfigOverwrite) -> Overwrite {
    if silent {
        Overwrite::Skip
    } else {
        match configured {
            ConfigOverwrite::Replace => Overwrite::Replace,
            ConfigOverwrite::Skip | ConfigOverwrite::Ask => Overwrite::Skip,
        }
    }
}

pub fn overwrite_wire(overwrite: Overwrite) -> &'static str {
    match overwrite {
        Overwrite::Skip => "skip",
        Overwrite::Replace => "replace",
    }
}

pub fn resolve_extract_dest(
    action: &LaunchAction,
    archive: &str,
    picked_dir: Option<&str>,
) -> Result<String> {
    match action {
        LaunchAction::ExtractHere => extract_here_dest(archive),
        LaunchAction::ExtractTo {
            dest_dir: Some(dir),
        } => {
            refuse_archive_as_dest(dir, archive)?;
            Ok(dir.clone())
        }
        LaunchAction::ExtractTo { dest_dir: None } => {
            let dest = picked_dir.ok_or_else(|| {
                ApiError::internal("extract-to destination omitted; folder picker required")
            })?;
            refuse_archive_as_dest(dest, archive)?;
            Ok(dest.to_string())
        }
        LaunchAction::Open | LaunchAction::IndexOnly => Err(ApiError::internal(
            "resolve_extract_dest called for a non-extract action",
        )),
    }
}

fn extract_here_dest(archive: &str) -> Result<String> {
    let path = Path::new(archive);
    match path.parent() {
        Some(parent) if !parent.as_os_str().is_empty() => Ok(parent.to_string_lossy().into_owned()),
        _ => Ok(".".to_string()),
    }
}

fn refuse_archive_as_dest(dest: &str, archive: &str) -> Result<()> {
    if same_path(dest, archive) {
        return Err(ApiError::internal(
            "extract-to dest omitted; archive path is not destDir",
        ));
    }
    Ok(())
}

fn same_path(a: &str, b: &str) -> bool {
    Path::new(a) == Path::new(b) || a.replace('\\', "/") == b.replace('\\', "/")
}
